### PR TITLE
firefox extension integration

### PR DESCRIPTION
**Pull request by Aman9das**
I'm trying to create an extension which will create an entry in context menus when right clicking a link..
To do:

- [x] Create extension
- [x] Create nativeMessaging port from extension side
- [x] Create manifest from application side
- [x] Create stdin and stdout handling in the application for nativeMessaging
- [x] Launch FireDM with the link recieved
- [ ] Move native manifest to the correct position depending on the system using a script

I need help at this.

Helpful resource: https://github.com/mdn/webextensions-examples/tree/master/native-messaging

image
Screenshot from 2021-11-13 21-26-12

### CODE REVIEW

**Comment by firedm** [comment]
@Aman9das 
nice work so far,
regarding native messaging it relys on stdin / stdout which will be a problem for our application, since gui applications on windows doesn't have stdin/stdout e.g. "frozen exe on windows prepared by cx_freeze"

to solve this we can rely on operating system clipboard, the web extension job is just copying the link to clipboard and FireDM automatically will catch the link and start processing

later we can improve the functionality based on special messages passed to clipboard

let me know if this is ok with you or not

by the way, there is a typo in context menu, should be FireDM instead of FreeDM

**Comment by Aman9das** [comment]
Please check if the extension is able to ping the app and back..

I saw the problem you stated, and so i just made browser.py to handle stdin and stdout, the url is stored in receivedMessage variable.

I hope this works :)

**Comment by firedm** [comment]
@Aman9das 
Hello Aman
will test it and let you know the result

**Comment by Aman9das** [comment]
Yes.. Note that you'll have to change the directory address in firedm.json to point to browser.py, and the icon.

**Comment by Aman9das** [comment]
OK could you tell how to launch firedm with a specific link?

**Comment by firedm** [comment]
> Did you test it out?, is it able to connect back and forth?

sorry for delay,
just start testing now and it gives this error: `this.trackerList is undefined  in  background.js:524 `

will try to test on windows and see if it works

**Comment by firedm** [comment]
> uget extension is quite helpful to learn how to support multiple browsers. Let's support get it running with firefox first, and then add support for chromium.

agree, lets make it running first with Firefox

> I could move the PR from master to another branch and we could try develop from there..

we could create branch name dev for this

still testing and will update with the results

**Comment by firedm** [comment]
> OK could you tell how to launch firedm with a specific link?

not applicable yet for gui to start with a link, however we can add it later

**Comment by Aman9das** [comment]
```this.trackerList is undefined in background.js:524```

That isnt related to the extension, but more with the browser configuration.. It certainly is some other background.js, mine doesnt have 524 lines. haha.

I also get ```sender.tab is undefined``` error in my browser, for example, without the fault of extension.

do you have matrix account? we can troubleshoot because setting up is trouble right now..

**Comment by firedm** [comment]
@Aman9das 
i spent some time today try to run addon on windows, sadly firefox keep giving me these messages
`The Web Console logging API (console.log, console.info, console.warn, console.error) has been disabled by a script on this page.`
tried to refresh Firefox, nothing changed
removed firefox and installed again, still getting same error
really don't know how to get rid of it



 
image


apart from this, say we managed to get messages passed thru stdin/stdout between addon and "browser.py" script
how we can solve the problem i mentioned earlier of gui application (frozen win32 [more..](https://stackoverflow.com/questions/54536/win32-gui-app-that-writes-usage-text-to-stdout-when-invoked-as-app-exe-help)) doesn't have sys.stdin / sys.stdout
since the script you offer "browser.py" need to be launched by firedm, and it will not have sys.stdin / sys.stdout available too, like parent process

what about using[ clipboard API's](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions) (clipboardRead, and clipboardWrite) to communicate between addon and FireDM

what do you think?


> Is there a matrix or xmpp room for development of firedm? we can troubleshoot there..

unfortunately, nothing set

**Comment by firedm** [comment]
@Aman9das 
created chat room /discussion space [here](https://github.com/firedm/FireDM/discussions/356) to continue our progress on this topic

**Comment by firedm** [comment]
@Aman9das 
to be able to merge this PR please remove all "nativeMessaging", since we use clipboard, 
after that you can add "nativeMessaging" in a future PR if it works as expected

File: add-on/background.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+
+/*
+On startup, connect to the "ping_pong" app.
+*/
+var port = browser.runtime.connectNative("firedm");
```
Comment (firedm) [inline add-on/background.js:5]: please remove this part from line 2

File: add-on/background.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+browser.menus.create({
+    id: "firedm",
+    title: browser.i18n.getMessage("menuItemFireDM"),
+    contexts: ["link"]
+});
+
+/*
+Listen for messages from the app.
+*/
+port.onMessage.addListener((response) => {
```
Comment (firedm) [inline add-on/background.js:47]: to be removed

File: add-on/background.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+});
+
+/*
+The click event listener, where we perform the appropriate action given the
+ID of the menu item that was clicked.
+*/
+browser.menus.onClicked.addListener((info) => {
+    if (info.menuItemId == "firedm") {
+    	  navigator.clipboard.writeText(info.linkUrl);
+          console.log(`Sent: ${info.linkUrl}`);
```
Comment (firedm) [inline add-on/background.js:58]: line 58, 59 to be removed

File: add-on/manifest.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  "icons": {
+    "48": "icons/firedm-48.png"
+  },
+  "background": {
+    "scripts": [
+      "background.js"
+    ]
+  },
+  "permissions": [
+    "nativeMessaging",
```
Comment (firedm) [inline add-on/manifest.json:22]: line 22 to be removed

File: firedm/browser.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env python3
```
Comment (firedm) [inline firedm/browser.py:1]: browser.py file to be removed

**Pull request closed without merging**